core/node/blocksync_request.cpp

```diff
@@ -0,0 +1,606 @@
+        if (!header_found) {
+          reduced.push_back(std::move(cid));
+        } else {
+          blocks_available.push_back(std::move(header_found.value()));
+        }
+      }
+      return reduced;
+    }
+
+    /// Stores part of blocksync response (i.e. tipset bundle)
+    outcome::result<void> storeTipsetBundle(
+        Ipld &ipld,
+        TipsetBundle &bundle,
+        bool store_messages,
+        const std::function<void(CID, BlockHeader)> &block_stored) {
+      size_t sz = bundle.blocks.size();
+
+      auto _msgs{bundle.messages};
+      if (_msgs) {
+        trace(
+            "storing tipset bundle of {} blocks, {} bls messages, {} secp "
+            "messages",
+            sz,
+            _msgs->bls_msgs.size(),
+            _msgs->secp_msgs.size());
+      } else {
+        trace("storing tipset bundle of {} blocks", sz);
+      }
+
+      std::vector<CID> secp_cids;
+      std::vector<CID> bls_cids;
+
+      try {
+        if (_msgs) {
+          if (_msgs->secp_msg_includes.size() != sz
+              || _msgs->bls_msg_includes.size() != sz) {
+            throw std::system_error(
+                BlocksyncRequest::Error::BLOCKSYNC_INCONSISTENT_RESPONSE);
+          }
+
+          secp_cids.reserve(_msgs->secp_msgs.size());
+          for (const auto &msg : _msgs->secp_msgs) {
+            OUTCOME_EXCEPT(cid, ipld.setCbor(msg));
+            secp_cids.push_back(std::move(cid));
+          }
+
+          bls_cids.reserve(_msgs->bls_msgs.size());
+          for (const auto &msg : _msgs->bls_msgs) {
+            OUTCOME_EXCEPT(cid, ipld.setCbor(msg));
+            bls_cids.push_back(std::move(cid));
+          }
+        }
+      } catch (const std::system_error &e) {
+        log()->error("cannot store tipset bundle, {}", e.code().message());
+        return e.code();
+      }
+
+      for (size_t i = 0; i < sz; ++i) {
+        auto &header{bundle.blocks[i]};
+        OUTCOME_TRY(block_cid, ipld.setCbor<BlockHeader>(header));
+        if (_msgs) {
+          MsgMeta meta;
+          ipld.load(meta);
+          for (auto idx : _msgs->secp_msg_includes[i]) {
+            if (idx >= secp_cids.size()) {
+              return BlocksyncRequest::Error::BLOCKSYNC_INCONSISTENT_RESPONSE;
+            }
+            OUTCOME_TRY(meta.secp_messages.append(secp_cids[idx]));
+          }
+          for (auto idx : _msgs->bls_msg_includes[i]) {
+            if (idx >= bls_cids.size()) {
+              return BlocksyncRequest::Error::BLOCKSYNC_INCONSISTENT_RESPONSE;
+            }
+            OUTCOME_TRY(meta.bls_messages.append(bls_cids[idx]));
+          }
+          OUTCOME_TRY(meta_cid, ipld.setCbor(meta));
+          if (meta_cid != header.messages) {
+            return BlocksyncRequest::Error::BLOCKSYNC_STORE_ERROR_CIDS_MISMATCH;
+          }
+          block_stored(std::move(block_cid), std::move(header));
+        }
+      }
+
+      return outcome::success();
+    }
+
+    // XXX
+    static int xxx = 0;
```

Comment:
What is it?

Reply:
Logging reference count when objects are created and destoryed